In [134]:
import pandas as pd
import re
# Carga de la base de datos de rotten Tomatoes-movies-and-critic-reviews-dataset

movies_df = pd.read_csv(
    "/kaggle/input/rotten-tomatoes-movies-and-critic-reviews-dataset/rotten_tomatoes_movies.csv"
)
movies_df.head()

rotten_tomatoes_link  \
0                              m/0814255   
1                              m/0878835   
2                                   m/10   
3                 m/1000013-12_angry_men   
4  m/1000079-20000_leagues_under_the_sea   

                                         movie_title  \
0  Percy Jackson & the Olympians: The Lightning T...   
1                                        Please Give   
2                                                 10   
3                    12 Angry Men (Twelve Angry Men)   
4                       20,000 Leagues Under The Sea   

                                          movie_info  \
0  Always trouble-prone, the life of teenager Per...   
1  Kate (Catherine Keener) and her husband Alex (...   
2  A successful, middle-aged Hollywood songwriter...   
3  Following the closing arguments in a murder tr...   
4  In 1866, Professor Pierre M. Aronnax (Paul Luk...   

                                   critics_consensus content_rating  \
0  Though it may seem like just another Harry Pot...             PG   
1  Nicole Holofcener's newest might seem slight i...              R   
2  Blake Edwards' bawdy comedy may not score a pe...              R   
3  Sidney Lumet's feature debut is a superbly wri...             NR   
4  One of Disney's finest live-action adventures,...              G   

                                              genres          directors  \
0  Action & Adventure, Comedy, Drama, Science Fic...     Chris Columbus   
1                                             Comedy  Nicole Holofcener   
2                                    Comedy, Romance      Blake Edwards   
3                                    Classics, Drama       Sidney Lumet   
4           Action & Adventure, Drama, Kids & Family  Richard Fleischer   

                                      authors  \
0  Craig Titley, Chris Columbus, Rick Riordan   
1                           Nicole Holofcener   
2                               Blake Edwards   
3                               Reginald Rose   
4                                 Earl Felton   

                                              actors original_release_date  \
0  Logan Lerman, Brandon T. Jackson, Alexandra Da...            2010-02-12   
1  Catherine Keener, Amanda Peet, Oliver Platt, R...            2010-04-30   
2  Dudley Moore, Bo Derek, Julie Andrews, Robert ...            1979-10-05   
3  Martin Balsam, John Fiedler, Lee J. Cobb, E.G....            1957-04-13   
4  James Mason, Kirk Douglas, Paul Lukas, Peter L...            1954-01-01   

   ...      production_company  tomatometer_status tomatometer_rating  \
0  ...        20th Century Fox              Rotten               49.0   
1  ...  Sony Pictures Classics     Certified-Fresh               87.0   
2  ...             Waner Bros.               Fresh               67.0   
3  ...    Criterion Collection     Certified-Fresh              100.0   
4  ...                  Disney               Fresh               89.0   

  tomatometer_count  audience_status  audience_rating audience_count  \
0             149.0          Spilled             53.0       254421.0   
1             142.0          Upright             64.0        11574.0   
2              24.0          Spilled             53.0        14684.0   
3              54.0          Upright             97.0       105386.0   
4              27.0          Upright             74.0        68918.0   

   tomatometer_top_critics_count  tomatometer_fresh_critics_count  \
0                             43                               73   
1                             44                              123   
2                              2                               16   
3                              6                               54   
4                              5                               24   

   tomatometer_rotten_critics_count  
0                                76  
1                                19  
2                                 8  


In [135]:
# Carga de la base de datos de rotten-tomatoes-movies-and-critic-reviews-dataset
reviews_df = pd.read_csv(
    "/kaggle/input/rotten-tomatoes-movies-and-critic-reviews-dataset/rotten_tomatoes_critic_reviews.csv"
)
reviews_df.head()

rotten_tomatoes_link      critic_name  top_critic           publisher_name  \
0            m/0814255  Andrew L. Urban       False           Urban Cinefile   
1            m/0814255    Louise Keller       False           Urban Cinefile   
2            m/0814255              NaN       False      FILMINK (Australia)   
3            m/0814255     Ben McEachen       False  Sunday Mail (Australia)   
4            m/0814255      Ethan Alter        True       Hollywood Reporter   

  review_type review_score review_date  \
0       Fresh          NaN  2010-02-06   
1       Fresh          NaN  2010-02-06   
2       Fresh          NaN  2010-02-09   
3       Fresh        3.5/5  2010-02-09   
4      Rotten          NaN  2010-02-10   

                                      review_content  
0  A fantasy adventure that fuses Greek mythology...  
1  Uma Thurman as Medusa, the gorgon with a coiff...  
2  With a top-notch cast and dazzling special eff...  
3  Whether audiences will get behind The Lightnin...  
4  What's really lacking in The Lightning Thief i...

## Preprocesamiento de rotten_tomatoes_movies.csv.

In [136]:
# Concatena titulo, genero, sinopsis y conseso en una sola columna
movies_df["text"] = (
    movies_df["movie_title"].fillna("") + " " +
    movies_df["genres"].fillna("") + " " +
    movies_df["movie_info"].fillna("") + " " +
    movies_df["critics_consensus"].fillna("")
)
movies_df["text"].head()

0    Percy Jackson & the Olympians: The Lightning T...
1    Please Give Comedy Kate (Catherine Keener) and...
2    10 Comedy, Romance A successful, middle-aged H...
3    12 Angry Men (Twelve Angry Men) Classics, Dram...
4    20,000 Leagues Under The Sea Action & Adventur...
Name: text, dtype: object

In [137]:
# Normaliza el texto a minuscilas, elimna caracteres especiales, limpia espacios
def normalize_text(text):
    text = text.lower()
    text = re.sub(r'[^a-z\s]', ' ', text)
    text = re.sub(r'\s+', ' ', text).strip()
    return text
movies_df["normalized_text"] = movies_df["text"].apply(normalize_text)
movies_df["normalized_text"].head()

0    percy jackson the olympians the lightning thie...
1    please give comedy kate catherine keener and h...
2    comedy romance a successful middle aged hollyw...
3    angry men twelve angry men classics drama foll...
4    leagues under the sea action adventure drama k...
Name: normalized_text, dtype: object

In [138]:
# Divide el texto en palabras individuales de 2 + letras
def tokenize(text):
    return re.findall(r'\b[a-z]{2,}\b', text)
movies_df["tokens"] = movies_df["normalized_text"].apply(tokenize)
movies_df["tokens"].head()

0    [percy, jackson, the, olympians, the, lightnin...
1    [please, give, comedy, kate, catherine, keener...
2    [comedy, romance, successful, middle, aged, ho...
3    [angry, men, twelve, angry, men, classics, dra...
4    [leagues, under, the, sea, action, adventure, ...
Name: tokens, dtype: object

In [139]:
# Elimina palabras comunes sin significado basado en el diccionario
STOPWORDS = {
    "the", "is", "in", "and", "to", "of", "a", "for", "on", "that", "it",
    "this", "as", "with", "by", "an", "are", "from", "at", "be", "was",
    "or", "which", "we", "can", "has", "have", "will", "not", "if", "but",
    "about", "there", "their", "they", "you", "your", "our", "my", "me",
    "all", "also", "so", "what", "when", "where", "how", "who", "do", "does",
    "did", "would", "could", "should", "may", "might", "must", "been", "being",
    "am", "i"
}
def remove_stopwords(tokens):
    return [word for word in tokens if word not in STOPWORDS]
movies_df["filtered_tokens"] = movies_df["tokens"].apply(remove_stopwords)
movies_df["filtered_tokens"].head()

0    [percy, jackson, olympians, lightning, thief, ...
1    [please, give, comedy, kate, catherine, keener...
2    [comedy, romance, successful, middle, aged, ho...
3    [angry, men, twelve, angry, men, classics, dra...
4    [leagues, under, sea, action, adventure, drama...
Name: filtered_tokens, dtype: object

In [140]:
# Reduce palabras a su raiz
def simple_stem(word):
    word = re.sub(r'(ss|ies|s)$', '', word)
    word = re.sub(r'(ing|ed)$', '', word)
    word = re.sub(r'(er|est)$', '', word)
    word = re.sub(r'ly$', '', word)
    return word

def stem_tokens(tokens):
    return [simple_stem(token) for token in tokens]

movies_df["stemmed_tokens"] = movies_df["filtered_tokens"].apply(stem_tokens)
movies_df["stemmed_tokens"].head()


0    [percy, jackson, olympian, lightn, thief, acti...
1    [please, give, comedy, kate, catherine, keen, ...
2    [comedy, romance, successful, middle, ag, holl...
3    [angry, men, twelve, angry, men, classic, dram...
4    [league, und, sea, action, adventure, drama, k...
Name: stemmed_tokens, dtype: object

In [141]:
# Funcion que ejecuta todo las funciones 
def preprocess(text):
    normalized = normalize_text(text)
    tokens = tokenize(normalized)
    filtered = remove_stopwords(tokens)
    stemmed = stem_tokens(filtered)
    return stemmed

movies_df["processed_text"] = movies_df["text"].apply(preprocess)
movies_df["processed_text"].head()


0    [percy, jackson, olympian, lightn, thief, acti...
1    [please, give, comedy, kate, catherine, keen, ...
2    [comedy, romance, successful, middle, ag, holl...
3    [angry, men, twelve, angry, men, classic, dram...
4    [league, und, sea, action, adventure, drama, k...
Name: processed_text, dtype: object

In [142]:
# Crear el dataset listo para TF-IDF
corpus = movies_df["processed_text"].tolist()
len(corpus)

17712

In [143]:
# Muestra un ejemplo de preporcesamiento
idx = 0
print("TITULO:")
print(movies_df.loc[idx, "movie_title"])
print("\nTEXTO PREPROCESADO (tokens):")
print(movies_df.loc[idx, "processed_text"])


TITULO:
Percy Jackson & the Olympians: The Lightning Thief

TEXTO PREPROCESADO (tokens):
['percy', 'jackson', 'olympian', 'lightn', 'thief', 'action', 'adventure', 'comedy', 'drama', 'science', 'fiction', 'fantasy', 'alway', 'trouble', 'prone', 'life', 'teenag', 'percy', 'jackson', 'logan', 'lerman', 'get', 'lot', 'more', 'complicat', 'he', 'learn', 'he', 'son', 'greek', 'god', 'poseidon', 'train', 'ground', 'children', 'deit', 'percy', 'learn', 'harne', 'hi', 'divine', 'pow', 'prepare', 'adventure', 'lifetime', 'he', 'prevent', 'feud', 'among', 'olympian', 'erupt', 'into', 'devastat', 'war', 'earth', 'rescue', 'hi', 'moth', 'clutche', 'hade', 'god', 'underworld', 'though', 'seem', 'like', 'just', 'anoth', 'harry', 'pott', 'knockoff', 'percy', 'jackson', 'benefit', 'strong', 'support', 'cast', 'speedy', 'plot', 'plenty', 'fun', 'greek', 'mythology']


## Preprocesamiento de rotten_tomatoes_critic_reviews.csv

In [144]:
# Solo selecciona columnas necesarias de critic_reviews
reviews_df = reviews_df[[
    "rotten_tomatoes_link",
    "review_type",
    "review_content"
]]
reviews_df.head()

rotten_tomatoes_link review_type  \
0            m/0814255       Fresh   
1            m/0814255       Fresh   
2            m/0814255       Fresh   
3            m/0814255       Fresh   
4            m/0814255      Rotten   

                                      review_content  
0  A fantasy adventure that fuses Greek mythology...  
1  Uma Thurman as Medusa, the gorgon with a coiff...  
2  With a top-notch cast and dazzling special eff...  
3  Whether audiences will get behind The Lightnin...  
4  What's really lacking in The Lightning Thief i...

In [145]:
# Elimina filas donde la review este vacio
reviews_df = reviews_df.dropna(subset=["review_content"])
reviews_df.shape

(1064211, 3)

In [146]:
# Procesa todo el contenido usando la funcion completa ya hecha
reviews_df["processed_text"] = reviews_df["review_content"].apply(preprocess)

In [147]:
# Agrupa las reviews de cada pelicula y las concatena en una sola lista de tokens
grouped_reviews = (
    reviews_df
    .groupby("rotten_tomatoes_link")["processed_text"]
    .sum()
    .reset_index()
)
grouped_reviews.head()

rotten_tomatoes_link  \
0                     m/+_one_2019   
1                             m/+h   
2                          m/-_man   
3  m/-cule_valley_of_the_lost_ants   
4                        m/0814255   

                                      processed_text  
0  [more, than, get, strength, it, fundamental, l...  
1  [ultimate, plush, exce, it, very, limit, expec...  
2  [too, many, piece, puzzle, left, out, come, of...  
3  [dialogue, free, bug, saga, carri, along, bril...  
4  [fantasy, adventure, fuse, greek, mythology, c...

In [148]:
# Unir movies con las reviews agrupadas
# Left join para mantener todas las películas
combined_df = movies_df.merge(
    grouped_reviews[['rotten_tomatoes_link', 'processed_text']], 
    on='rotten_tomatoes_link', 
    how='left',
    suffixes=('_movie', '_reviews')
)

print(f"Peliculas originales: {len(movies_df)}")
print(f"Peliculas con reviews: {len(grouped_reviews)}")
print(f"Dataset combinado: {len(combined_df)}")

Peliculas originales: 17712
Peliculas con reviews: 17697
Dataset combinado: 17712


In [149]:
# Combina tokens procesados de la pelicula con los de las reviews
def combine_tokens(row):
    movie_tokens = row['processed_text_movie'] if isinstance(row['processed_text_movie'], list) else []
    review_tokens = row['processed_text_reviews'] if isinstance(row['processed_text_reviews'], list) else []
    return movie_tokens + review_tokens

combined_df['all_processed_text'] = combined_df.apply(combine_tokens, axis=1)

# Ver estadísticas
combined_df['total_tokens'] = combined_df['all_processed_text'].apply(len)
print("Estadisticas texto combinado:")
print(combined_df['total_tokens'].describe())

Estadisticas texto combinado:
count    17712.000000
mean       863.295111
std       1176.344573
min          2.000000
25%        170.000000
50%        384.000000
75%       1071.000000
max      15524.000000
Name: total_tokens, dtype: float64


In [150]:
# Ejemplo
idx = 0
print("=" * 80)
print("PELÍCULA DE EJEMPLO:")
print("=" * 80)
print(f"\nTU¿ITULO: {combined_df.loc[idx, 'movie_title']}")
print(f"\nTOKENS PELICULA: {len(combined_df.loc[idx, 'processed_text_movie'])}")
print(combined_df.loc[idx, 'processed_text_movie'][:20])

if isinstance(combined_df.loc[idx, 'processed_text_reviews'], list):
    print(f"\nTOKENS REVIEWS: {len(combined_df.loc[idx, 'processed_text_reviews'])}")
    print(combined_df.loc[idx, 'processed_text_reviews'][:20])
else:
    print("\nNo tiene reviews")

PELÍCULA DE EJEMPLO:

TU¿ITULO: Percy Jackson & the Olympians: The Lightning Thief

TOKENS PELICULA: 82
['percy', 'jackson', 'olympian', 'lightn', 'thief', 'action', 'adventure', 'comedy', 'drama', 'science', 'fiction', 'fantasy', 'alway', 'trouble', 'prone', 'life', 'teenag', 'percy', 'jackson', 'logan']

TOKENS REVIEWS: 2066
['fantasy', 'adventure', 'fuse', 'greek', 'mythology', 'contemporary', 'american', 'place', 'value', 'anyone', 'around', 'give', 'take', 'couple', 'year', 'thrill', 'visual', 'spectacle', 'uma', 'thurman']


In [151]:
# Guardar las columnas importantes para usarlo en TF-IF
final_columns = [
    'movie_title', 
    'rotten_tomatoes_link',
    'processed_text_movie',
    'processed_text_reviews', 
    'all_processed_text'
]
optional_columns = ['genres', 'tomatometer_rating', 'audience_rating', 'original_release_date']
for col in optional_columns:
    if col in combined_df.columns:
        final_columns.insert(1, col)

processed_df = combined_df[final_columns].copy()
processed_df.head()

movie_title original_release_date  \
0  Percy Jackson & the Olympians: The Lightning T...            2010-02-12   
1                                        Please Give            2010-04-30   
2                                                 10            1979-10-05   
3                    12 Angry Men (Twelve Angry Men)            1957-04-13   
4                       20,000 Leagues Under The Sea            1954-01-01   

   audience_rating  tomatometer_rating  \
0             53.0                49.0   
1             64.0                87.0   
2             53.0                67.0   
3             97.0               100.0   
4             74.0                89.0   

                                              genres  \
0  Action & Adventure, Comedy, Drama, Science Fic...   
1                                             Comedy   
2                                    Comedy, Romance   
3                                    Classics, Drama   
4           Action & Adventure, Drama, Kids & Family   

                    rotten_tomatoes_link  \
0                              m/0814255   
1                              m/0878835   
2                                   m/10   
3                 m/1000013-12_angry_men   
4  m/1000079-20000_leagues_under_the_sea   

                                processed_text_movie  \
0  [percy, jackson, olympian, lightn, thief, acti...   
1  [please, give, comedy, kate, catherine, keen, ...   
2  [comedy, romance, successful, middle, ag, holl...   
3  [angry, men, twelve, angry, men, classic, dram...   
4  [league, und, sea, action, adventure, drama, k...   

                              processed_text_reviews  \
0  [fantasy, adventure, fuse, greek, mythology, c...   
1  [like, holofcen, previou, picture, please, giv...   
2  [known, it, numerical, rat, system, beautiful,...   
3  [film, texture, humour, relevance, time, minut...   
4  [embodiment, disney, hi, b, fami, entertainmen...   

                                  all_processed_text  
0  [percy, jackson, olympian, lightn, thief, acti...  
1  [please, give, comedy, kate, catherine, keen, ...  
2  [comedy, romance, successful, middle, ag, holl...  
3  [angry, men, twelve, angry, men, classic, dram...  
4  [league, und, sea, action, adventure, drama, k...

## Construcción del Sistema

In [152]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import numpy as np

# Convertir listas de tokens a strings
processed_df['text_for_tfidf'] = processed_df['all_processed_text'].apply(lambda x: ' '.join(x))

# Crear el vectorizador TF-IDF
tfidf_vectorizer = TfidfVectorizer(
    max_features=5000,  # Limitar a las 5000 palabras más importantes
    min_df=2,           # Una palabra debe aparecer en al menos 2 documentos
    max_df=0.8          # Ignorar palabras que aparecen en más del 80% de documentos
)

# Crear la matriz TF-IDF
tfidf_matrix = tfidf_vectorizer.fit_transform(processed_df['text_for_tfidf'])
print(f"Matriz TF-IDF creada:")
print(f"Dimensiones: {tfidf_matrix.shape}")
print(f"Películas: {tfidf_matrix.shape[0]}")

Matriz TF-IDF creada:
Dimensiones: (17712, 5000)
Películas: 17712


## Consulta


In [158]:
# Busca una pelicula relevante para una consulta
# Devuelve un  DataFrame con los resultados más relevantes
def search_movies(query, top_k=10, show_details=True):

    # Preprocesa la consulta 
    query_tokens = preprocess(query)
    query_text = ' '.join(query_tokens)
    # Vectoriza la consulta
    query_vector = tfidf_vectorizer.transform([query_text])
    # Calcula similitud coseno con todas las peliculas
    similarities = cosine_similarity(query_vector, tfidf_matrix).flatten()
    # Obtener indices de peliculas más similares
    top_indices = similarities.argsort()[-top_k:][::-1]
    # Crear DataFrame con resultados
    results = processed_df.iloc[top_indices].copy()
    results['similarity_score'] = similarities[top_indices]
    results['rank'] = range(1, top_k + 1)
    if show_details:
        print(f"CONSULTA: '{query}'")
        print(f"Tokens procesados: {query_tokens}")
        
        for idx, row in results.iterrows():
            print(f"\n{row['rank']}. {row['movie_title']}")
            print(f"   Similitud: {row['similarity_score']:.4f}")
            if 'genres' in row:
                print(f"   Generos: {row['genres']}")
            if 'tomatometer_rating' in row:
                print(f"   Rating críticos: {row['tomatometer_rating']}")
            if 'audience_rating' in row:
                print(f"   Rating audiencia: {row['audience_rating']}")
            print("--------------------")
    
    return results[['rank', 'movie_title', 'similarity_score', 'genres', 
                    'tomatometer_rating', 'audience_rating', 'original_release_date']]

In [157]:
# Ejemplo de consulta
results_space = search_movies("Love Strange", top_k=10)

CONSULTA: 'Love Strange'
Tokens procesados: ['love', 'strange']

1. Love Is Strange
   Similitud: 0.6995
   Géneros: Drama, Special Interest
   Rating críticos: 93.0
   Rating audiencia: 66.0
----------------------------------------------------------------------------------------------------

2. Strange Magic
   Similitud: 0.3923
   Géneros: Animation
   Rating críticos: 16.0
   Rating audiencia: 52.0
----------------------------------------------------------------------------------------------------

3. Doctor Strange
   Similitud: 0.3603
   Géneros: Action & Adventure, Science Fiction & Fantasy
   Rating críticos: 89.0
   Rating audiencia: 86.0
----------------------------------------------------------------------------------------------------

4. The Strange Ones
   Similitud: 0.3557
   Géneros: Drama, Mystery & Suspense
   Rating críticos: 55.0
   Rating audiencia: 43.0
----------------------------------------------------------------------------------------------------

5. Strange 

## Conclusiones

En la consulta “Love Strange”, el sistema funciono bien porque la pelicula Love Is Strange apareció en el primer lugar con una similitud alta de 0.69, lo que indica una gran coincidencia entre los terminos de la consulta y el contenido de la pelicula. Los siguientes resultados, como Strange Magic (0.39) y Doctor Strange (0.36), tienen similitudes menores porque solo comparten una de las palabras clave o están relacionadas de forma parcial. Esto muestra que el uso de TF-IDF junto con el preprocesamiento permitió diferenciar correctamente las peliculas mas relevantes de las menos relacionadas, generando un ranking coherente y alineado con la consulta realizada.